presto-main/src/main/java/com/facebook/presto/operator/aggregation/ArrayAggregation.java

```diff
@@ -0,0 +1,149 @@
+    private static final Method COMBINE_FUNCTION = method(ArrayAggregation.class, "combine", ArrayAggregationState.class, ArrayAggregationState.class);
+    private static final Signature SIGNATURE = new Signature(NAME, ImmutableList.of(typeParameter("T")), "array<T>", ImmutableList.of("T"), false, false);
+
+    @Override
+    public Signature getSignature()
+    {
+        return SIGNATURE;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "return an array of values";
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type valueType = types.get("T");
+        Signature signature = new Signature(NAME,
+                parameterizedTypeName("array", valueType.getTypeSignature()),
+                valueType.getTypeSignature());
+        InternalAggregationFunction aggregation = generateAggregation(valueType);
+        return new FunctionInfo(signature, getDescription(), aggregation.getIntermediateType().getTypeSignature(), aggregation, false);
+    }
+
+    private static InternalAggregationFunction generateAggregation(Type valueType)
+    {
+        DynamicClassLoader classLoader = new DynamicClassLoader(ArrayAggregation.class.getClassLoader());
+
+        ArrayAggregationStateSerializer stateSerializer = new ArrayAggregationStateSerializer();
+        Type intermediateType = stateSerializer.getSerializedType();
+
+        List<Type> inputTypes = ImmutableList.of(valueType);
+        Type outputType = new ArrayType(valueType);
+        ArrayAggregationStateFactory stateFactory = new ArrayAggregationStateFactory(new ArrayType(valueType));
+        AggregationMetadata metadata = new AggregationMetadata(
+                generateAggregationName(NAME, valueType, inputTypes),
+                createInputParameterMetadata(valueType),
+                INPUT_FUNCTION,
+                null,
+                null,
+                COMBINE_FUNCTION,
+                OUTPUT_FUNCTION,
+                ArrayAggregationState.class,
+                stateSerializer,
+                stateFactory,
+                outputType,
+                false);
+
+        GenericAccumulatorFactoryBinder factory = new AccumulatorCompiler().generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(NAME, inputTypes, intermediateType, valueType, true, false, factory);
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type value)
+    {
+        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(INPUT_CHANNEL, value), new ParameterMetadata(BLOCK_INDEX));
+    }
+
+    public static void input(ArrayAggregationState state, Block value, int position)
+    {
+        Type t = state.getArrayType().getElementType();
+        Object v = t.getObjectValue(null /* session */, value, position);
```

Comment:
Is it better to keep an array of blocks instead of actual objects? Going to try that and run some perf experiments. Right now, the perf doesn't look very good.